crates/ibc/src/core/ics24_host/identifier.rs

```diff
@@ -123,46 +101,89 @@ impl ChainId {
+    /// assert_eq!(ChainId::new("chainA", 1).with_version(2), ChainId::new("chainA", 2));
     /// assert_eq!("chain1".parse::<ChainId>().unwrap().with_version(2), "chain1".parse::<ChainId>().unwrap());
     /// ```
     pub fn with_version(mut self, version: u64) -> Self {
-        if Self::is_epoch_format(&self.id) {
-            self.id = {
-                let mut split: Vec<&str> = self.id.split('-').collect();
-                let version = version.to_string();
-                if let Some(last_elem) = split.last_mut() {
-                    *last_elem = &version;
-                }
-                split.join("-")
-            };
-            self.version = version;
+        if self.version != 0 {
+            if let (name, Some(_)) = Self::split_chain_id(&self.id) {
+                self.id = format!("{name}-{version}");
+                self.version = version;
+            }
         }
         self
     }
+
+    /// Splits chain_id into name and version if the id includes epoch number.
+    ///
+    /// Chain id with epoch number has format `{name}-{version}` where version
+    /// is a positive integer starting with non-zero digit.  If `chain_id` is in
+    /// that format, returns `(name, Some(version))`; otherwise returns
+    /// `(chain_id, None)`.
+    ///
+    /// **Panics** if chain id is in epoch format but the version overflows
+    /// `u64`.
+    fn split_chain_id(chain_id: &str) -> (&str, Option<u64>) {
+        fn split(chain_id: &str) -> Option<(&str, u64)> {
+            let (name, version) = chain_id.rsplit_once('-')?;
+            if name.is_empty() {
+                return None;
+            }
+            let (car, cdr) = version.as_bytes().split_first()?;
+            if b'1' <= *car && *car <= b'9' && cdr.iter().all(u8::is_ascii_digit) {
+                Some((name, version.parse().unwrap()))
+            } else {
+                None
+            }
+        }
+        match split(chain_id) {
+            Some((name, version)) => (name, Some(version)),
+            None => (chain_id, None)
+        }
+    }
+}
+
+impl Default for ChainId {
+    fn default() -> Self {
+        Self { id: String::from(DEFAULT_CHAIN_ID), version: 0 }
+    }
+}
+
+impl<'a> From<&'a str> for ChainId {
+    fn from(value: &'a str) -> Self {
+        let version = Self::chain_version(value);
+        Self { id: String::from(value), version }
+    }
+}
+
+impl From<String> for ChainId {
+    fn from(value: String) -> Self {
+        let version = Self::chain_version(&value);
+        Self { id: value, version }
+    }
 }
 
 impl FromStr for ChainId {
     type Err = Infallible;
 
     fn from_str(id: &str) -> Result<Self, Self::Err> {
-        Ok(Self::from_string(id))
+        Ok(id.into())
     }
 }
 
-impl Display for ChainId {
+impl core::fmt::Display for ChainId {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
-        write!(f, "{}", self.id)
+        self.id.fmt(f)
```

Comment:
Why do this?

Reply:
Honestly, I don’t trust Rust’s ability to optimise `write!(f, "{}", self.id)`.  I haven’t looked closely at this, but for example `foo.to_string()` is faster than `format!("{}", foo)` and I suspect the same is true in this case.